File: tools/portability/package-manager-sodium-gate.mjs
#!/usr/bin/env node
// package-manager-sodium-gate.mjs — gate E2E do wiring REAL entre o contrato
// IPackageManager (§6 item 5) e o libsodium (ed25519): compila o gate C++
// (package-manager-sodium-gate.cpp) contra o adapter real
// (src/engine/sdk/PackageManager.cpp) + sodium-verifier.hpp + a lib estática
// MSVC, e roda o fluxo completo (verify/install/tamper/wrong-key). Exit 0 =
// wiring real funcionando. Step 7f6b no ci-matrix.
import { spawnSync } from 'child_process';
import { existsSync, writeFileSync } from 'fs';
import { join } from 'path';

const ROOT = process.cwd();
const SODIUM = join(ROOT, 'external', 'solutions', 'libsodium');
const INCLUDE_SODIUM = join(SODIUM, 'src', 'libsodium', 'include');
const LIB = join(SODIUM, 'bin', 'x64', 'Release', 'v143', 'static', 'libsodium.lib');
const EXE = join(ROOT, 'build', 'Release', 'package-manager-sodium-gate.exe');

if (!existsSync(LIB)) {
  console.error('[package-manager-sodium-gate] FAIL: libsodium lib missing — run libsodium-gate first');
  process.exit(1);
}

const vcvars = 'C:\\Program Files\\Microsoft Visual Studio\\18\\Community\\VC\\Auxiliary\\Build\\vcvars64.bat';
const w = (p) => p.replace(/\//g, '\\');
const bat = [
  '@echo off',
  `call "${vcvars}" >nul 2>&1`,
  `cd /d ${w(ROOT)}`,
  `cl /nologo /EHsc /std:c++17 /O1 /DSODIUM_STATIC /I ${w(INCLUDE_SODIUM)} /I ${w(join(ROOT, 'src', 'engine', 'public'))} /I ${w(join(ROOT, 'src'))} /I ${w(join(ROOT, 'tools', 'portability'))} ${w(join(ROOT, 'tools', 'portability', 'package-manager-sodium-gate.cpp'))} ${w(join(ROOT, 'src', 'engine', 'sdk', 'PackageManager.cpp'))} ${w(LIB)} /Fe:${w(EXE)}`,
  'if errorlevel 1 exit /b 1',
  w(EXE)
].join('\r\n');
const batPath = join(ROOT, 'tools', 'portability', 'build-package-manager-sodium.bat');
writeFileSync(batPath, bat, 'utf8');

const run = spawnSync('cmd.exe', ['/c', batPath], { encoding: 'utf8' });
if (run.status !== 0) {
  console.error('[package-manager-sodium-gate] FAIL');
  console.error(run.stdout + run.stderr);
  process.exit(1);
}
process.stdout.write(run.stdout);
console.log('[package-manager-sodium-gate] PASS — IPackageManager + libsodium ed25519 wiring real');
process.exit(0);

File: tools/portability/package-manager-sodium-gate.cpp
// package-manager-sodium-gate.cpp — gate E2E do wiring REAL entre o contrato
// IPackageManager (§6 item 5) e o libsodium (ed25519) via o verifier concreto
// SodiumSignatureVerifier. Prova o fluxo completo que o finding #298
// destravou: manifestos registrados, assinaturas geradas com crypto_sign REAL
// por package (pk|sig), instalação com assinaturas válidas (deps verificadas
// explicitamente + root no install), assinatura corrompida recusada
// (all-or-nothing), chave errada recusada. Compila contra o IPackageManager
// real (src/engine/sdk/PackageManager.cpp) + libsodium estática MSVC.
// Exit 0 = wiring real funcionando.

#include "engine/packaging/IPackageManager.hpp"
#include "sodium-verifier.hpp"

#include <cstdio>
#include <string>
#include <vector>

namespace {

int failures = 0;

void check(bool ok, const char* what) {
    if (!ok) {
        std::printf("  FAIL: %s\n", what);
        ++failures;
    }
}

// Assina contentHash com ed25519 e devolve "hex(pk)|hex(sig)".
std::string sign(const std::string& contentHash, unsigned char* pk) {
    unsigned char sk[crypto_sign_SECRETKEYBYTES];
    crypto_sign_keypair(pk, sk);
    unsigned char sig[crypto_sign_BYTES];
    crypto_sign_detached(sig, nullptr, reinterpret_cast<const unsigned char*>(contentHash.data()),
                         contentHash.size(), sk);
    char buf[256];
    std::string out;
    for (int i = 0; i < crypto_sign_PUBLICKEYBYTES; ++i) {
        std::snprintf(buf, sizeof(buf), "%02x", pk[i]);
        out += buf;
    }
    out += '|';
    for (int i = 0; i < crypto_sign_BYTES; ++i) {
        std::snprintf(buf, sizeof(buf), "%02x", sig[i]);
        out += buf;
    }
    return out;
}

engine::packaging::PackageManifest pkg(const char* name, const char* version,
                                       const char* hash,
                                       std::initializer_list<const char*> deps) {
    engine::packaging::PackageManifest m;
    m.name = name;
    m.version = version;
    m.content_hash = hash;
    const std::vector<const char*> raw(deps);
    for (std::size_t i = 0; i + 1 < raw.size(); i += 2) {
        engine::packaging::PackageDependency d;
        d.name = raw[i];
        d.constraint = raw[i + 1];
        m.dependencies.push_back(d);
    }
    return m;
}

}  // namespace

int main() {
    std::printf("[package-manager + libsodium E2E]\n");
    if (sodium_init() < 0) {
        std::printf("FAIL: sodium_init\n");
        return 1;
    }

    std::string err;
    auto pm = engine::packaging::create_package_manager("pm-sodium", err);
    check(pm != nullptr, "package manager criado");
    engine::packaging::SodiumSignatureVerifier verifier;
    pm->set_verifier(&verifier);

    pm->register_manifest(pkg("core", "1.0", "hash-core", {}), err);
    pm->register_manifest(pkg("mod", "2.0", "hash-mod", {"core", ">=1.0"}), err);

    // Gera assinaturas reais (chaves distintas por package).
    unsigned char pk_core[crypto_sign_PUBLICKEYBYTES];
    unsigned char pk_mod[crypto_sign_PUBLICKEYBYTES];
    const std::string sig_core = sign("hash-core", pk_core);
    const std::string sig_mod = sign("hash-mod", pk_mod);

    // 1. Assinatura válida do root + dep verificada explicitamente → instala.
    check(pm->verify_signature("core", sig_core, err), "verify real de core ok");
    check(pm->install("mod", sig_mod, err), "install de mod com assinatura real ok");
    check(pm->installed().size() == 2, "core + mod instalados");

    // 2. Registra um segundo grafo e testa as recusas REAIS do crypto.
    pm->register_manifest(pkg("core2", "1.0", "hash-core2", {}), err);
    pm->register_manifest(pkg("evil", "1.0", "hash-evil", {"core2", "*"}), err);
    unsigned char pk_core2[crypto_sign_PUBLICKEYBYTES];
    const std::string sig_core2 = sign("hash-core2", pk_core2);
    const std::string sig_evil = sign("hash-evil", pk_mod);
    check(pm->verify_signature("core2", sig_core2, err), "verify real de core2 ok");

    // Tamper (um char no hex) → recusa no verify E no install, sem mutar.
    std::string sig_core2_bad = sig_core2;
    sig_core2_bad[0] = (sig_core2_bad[0] == '0') ? '1' : '0';
    check(!pm->verify_signature("core2", sig_core2_bad, err), "tamper rejeitado no verify");

    // Assinatura de OUTRO conteúdo (mensagem errada) → recusa no install.
    std::string sig_wrong_content = sign("hash-other", pk_core2);
    check(!pm->install("evil", sig_wrong_content, err),
          "assinatura de outro conteúdo recusada no install");

    // Malformada → recusa.
    check(!pm->install("evil", "tampered|deadbeef", err), "assinatura malformada recusada");
    check(pm->installed().size() == 2, "nada novo instalado (all-or-nothing)");

    // 3. Válida → instala core2 + evil (grafo novo completo). O verify
    //    fracassado acima invalidou core2 (signature_valid=false — correto),
    //    então re-verifica com a assinatura boa antes do install.
    check(pm->verify_signature("core2", sig_core2, err), "re-verify real de core2 ok");
    check(pm->install("evil", sig_evil, err), "install de evil com assinatura real ok");
    check(pm->installed().size() == 4, "4 instalados (core, mod, core2, evil)");

    if (failures == 0) {
        std::printf("package-manager-sodium-gate: ALL PASSED (IPackageManager + libsodium ed25519 wiring real)\n");
        return 0;
    }
    std::printf("package-manager-sodium-gate: %d FAILURE(S)\n", failures);
    return 1;
}
